dht_net.go

```diff
@@ -153,206 +154,3 @@ func (dht *IpfsDHT) updateFromMessage(ctx context.Context, p peer.ID, mes *pb.Me
 	}
 	return nil
 }
-
-func (dht *IpfsDHT) messageSenderForPeer(ctx context.Context, p peer.ID) (*messageSender, error) {
-	dht.smlk.Lock()
-	ms, ok := dht.strmap[p]
-	if ok {
-		dht.smlk.Unlock()
-		return ms, nil
-	}
-	ms = &messageSender{p: p, dht: dht}
-	dht.strmap[p] = ms
-	dht.smlk.Unlock()
-
-	if err := ms.prepOrInvalidate(ctx); err != nil {
-		dht.smlk.Lock()
-		defer dht.smlk.Unlock()
-
-		if msCur, ok := dht.strmap[p]; ok {
-			// Changed. Use the new one, old one is invalid and
-			// not in the map so we can just throw it away.
-			if ms != msCur {
-				return msCur, nil
-			}
-			// Not changed, remove the now invalid stream from the
-			// map.
-			delete(dht.strmap, p)
-		}
-		// Invalid but not in map. Must have been removed by a disconnect.
-		return nil, err
-	}
-	// All ready to go.
-	return ms, nil
-}
-
-type messageSender struct {
-	s   inet.Stream
-	r   ggio.ReadCloser
-	w   bufferedWriteCloser
-	lk  sync.Mutex
-	p   peer.ID
-	dht *IpfsDHT
-
-	invalid   bool
-	singleMes int
-}
-
-// invalidate is called before this messageSender is removed from the strmap.
-// It prevents the messageSender from being reused/reinitialized and then
-// forgotten (leaving the stream open).
-func (ms *messageSender) invalidate() {
-	ms.invalid = true
-	if ms.s != nil {
-		ms.s.Reset()
-		ms.s = nil
-	}
-}
-
-func (ms *messageSender) prepOrInvalidate(ctx context.Context) error {
-	ms.lk.Lock()
-	defer ms.lk.Unlock()
-	if err := ms.prep(ctx); err != nil {
-		ms.invalidate()
-		return err
-	}
-	return nil
-}
-
-func (ms *messageSender) prep(ctx context.Context) error {
-	if ms.invalid {
-		return fmt.Errorf("message sender has been invalidated")
-	}
-	if ms.s != nil {
-		return nil
-	}
-
-	nstr, err := ms.dht.host.NewStream(ctx, ms.p, ms.dht.protocols...)
-	if err != nil {
-		return err
-	}
-
-	ms.r = ggio.NewDelimitedReader(nstr, inet.MessageSizeMax)
-	ms.w = newBufferedDelimitedWriter(nstr)
-	ms.s = nstr
-
-	return nil
-}
-
-// streamReuseTries is the number of times we will try to reuse a stream to a
-// given peer before giving up and reverting to the old one-message-per-stream
-// behaviour.
-const streamReuseTries = 3
```

Comment:
So, we had this to support the old `/ipfs/dht` protocol (that only supported one message per stream. Honestly, we can probably drop support (our DHT crawlers aren't seeing any nodes that only support this old protocol) but, if we do, we should stop _advertising_ this protocol (remove it from the `DefaultProtocols` list.

Reply:
Okay thanks.

Reply:
I can add the deprecation here, or we can do it in another PR. I'll add the notice here for now. If I understand correctly, this stream pooling implementation will succeed on the first message on every new stream, and then fail on later messages: opening new streams to compensate (but not until a later request/send attempt). Does that sound correct?

Reply:
This patch doesn't:

1. Retry on failure.
2. Track these failures and fallback on using streams once.

That means the second of two back-to-back requests using the old protocol will always fail and we won't do anything to correct (retry, remember that the peer is using the old protocol, etc.). That's why the old message senders have all that complicated resend logic.

Reply:
Okay that's good to know. I believe implementing the retry logic in this implementation will be very easy. It's preferable to deprecate it entirely though.

Reply:
Let's just deprecate it. There's an edge case where the stream might close _just_ as we're about to use it but, meh.

Reply:
Do you mean rip the string out of the DefaultProtocols too?

Reply:
Yes.